concent_api/core/views.py

```diff
@@ -1,223 +1,321 @@
-            store_message(data['type'], data['message_cannot_commpute_task'], request.body)
+            store_message(message.__class__.__name__, message_task_to_compute, request.body)
             return HttpResponse("", status = 202)
+        else:
+            raise Http400("Time to acknowledge this task is already over.")
+    else:
+        if hasattr(message, 'TYPE'):
+            raise Http400("This message type ({}) is either not supported or cannot be submitted to Concent.".format(message.TYPE))
+        else:
+            raise Http400("Unknown message type or not a Golem message.")
 
 
 @api_view
 @require_POST
-def receive(_request, _message):
+def receive(request, _message):
     undelivered_message_statuses    = MessageStatus.objects.filter(delivered = False)
-    last_undelivered_message_status = undelivered_message_statuses.order_by('id').last()
-
+    last_undelivered_message_status = undelivered_message_statuses.order_by('timestamp').last()
     if last_undelivered_message_status is None:
         return None
 
-    # Mark message as delivered
-    last_undelivered_message_status.delivered = True
-    last_undelivered_message_status.save()
+    current_time = int(datetime.datetime.now().timestamp())
 
+    client_public_key = decode_client_public_key(request)
     raw_message_data     = last_undelivered_message_status.message.data.tobytes()
-    decoded_message_data = json.loads(raw_message_data.decode('utf-8'))
-    current_time         = int(datetime.datetime.now().timestamp())
+    decoded_message_data = load(
+        raw_message_data,
+        settings.CONCENT_PRIVATE_KEY,
+        client_public_key,
+    )
 
-    if decoded_message_data['type'] == "MessageAckReportComputedTask":
-        if current_time <= decoded_message_data['message_task_to_compute']['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME:
-            return decoded_message_data
+    # Mark message as delivered
+    if last_undelivered_message_status.message.type == "MessageForceReportComputedTask":
+        message_task_to_compute_from_force = load(
+            decoded_message_data.message_task_to_compute,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key)
+        if message_task_to_compute_from_force.deadline + settings.CONCENT_MESSAGING_TIME < current_time:
+            last_undelivered_message_status.delivered = True
+            last_undelivered_message_status.save()
+            return dump(
+                MessageAckReportComputedTask(
+                    message_task_to_compute = decoded_message_data.message_task_to_compute
+                ),
+                settings.CONCENT_PRIVATE_KEY,
+                client_public_key,
+            )
+    else:
+        last_undelivered_message_status.delivered = True
+        last_undelivered_message_status.save()
+
+    if isinstance(decoded_message_data, MessageForceReportComputedTask):
```

Comment:
This will be always false. You have already processed `MessageForceReportComputedTask` above.